plugins/inputs/prometheus/target_allocator.go

```diff
@@ -0,0 +1,154 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"github.com/aws/amazon-cloudwatch-agent/cfg/envconfig"
+	"os"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	otelpromreceiver "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver"
+	tamanager "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusreceiver/targetallocator"
+	promconfig "github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/discovery"
+	"github.com/prometheus/prometheus/scrape"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/confmap"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/receiver"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"gopkg.in/yaml.v3"
+)
+
+const DEFAULT_TLS_CA_FILE_PATH = "/etc/amazon-cloudwatch-observability-agent-cert/tls-ca.crt"
+
+type TargetAllocatorManager struct {
+	enabled   bool
+	manager   *tamanager.Manager
+	config    *otelpromreceiver.Config
+	host      component.Host
+	sm        *scrape.Manager
+	dm        *discovery.Manager
+	smLiveCh  chan struct{}
+	dmLiveCh  chan struct{}
+	taReadyCh chan struct{}
+}
+
+func isPodNameAvailable() bool {
+	podName := os.Getenv(envconfig.PodName)
+	if podName == "" {
+		return false
+	}
+	return true
+}
+func loadConfigFromFilename(filename string) (*otelpromreceiver.Config, error) {
+	yamlFile, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	var stringMap map[string]interface{}
+	err = yaml.Unmarshal(yamlFile, &stringMap)
+	if err != nil {
+		return nil, err
+	}
+	componentParser := confmap.NewFromStringMap(stringMap)
+	if componentParser == nil {
+		return nil, fmt.Errorf("unable to parse config from filename %s", filename)
+	}
+	var cfg otelpromreceiver.Config
+	err = componentParser.Unmarshal(&cfg)
+	if err != nil {
+		return nil, err
+	}
+	return &cfg, nil
+}
+
+// Adapter from go-kit/log to zap.Logger
+func createLogger(level zapcore.Level) *zap.Logger {
+	// Create a base zap logger (you can customize it as needed)
+	zapCore := zapcore.NewCore(
+		zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()), // Use JSON encoder for zap
+		zapcore.AddSync(os.Stdout),                               // Output to stdout
+		level,                                                    // Set log level to Debug
+	)
+	// Create the zap logger
+	zapLogger := zap.New(zapCore)
+	return zapLogger
+}
+
+func createTargetAllocatorManager(filename string, logger log.Logger, sm *scrape.Manager, dm *discovery.Manager, smDoneCh, dmDoneCh chan struct{}) *TargetAllocatorManager {
+	tam := TargetAllocatorManager{
+		enabled: false,
+		manager: nil,
+		config:  nil,
+		host:    nil,
+		sm:      sm,
+		dm:      dm,
+	}
+	tam.smLiveCh = smDoneCh
+	tam.dmLiveCh = dmDoneCh
```

Comment:
Shouldn't these both be initialized to `make(chan struct{}, 1)` here like `tam.taReadyCh` is for consistency?

Reply:
No because these channels are owned by other managers since it represents they being done. 
more information here: https://github.com/aws/amazon-cloudwatch-agent/pull/1394#discussion_r1805307867

Reply:
Cool, lgtm then.